src/sandbox.rs

```diff
@@ -0,0 +1,675 @@
+/// A minimalistic ptrace-based sandbox.
+///
+/// Modern (2025-era) sandboxes should use seccomp-bpf + user notification, but at Replit, there's already one such sandbox in use, so it cannot be used.
+/// Instead, an old (2000s-era) ptrace-based sandbox is used. It is not intended to be secure, just to prevent accidents.
+use std::fs::read_link;
+use std::panic::catch_unwind;
+use std::path::PathBuf;
+
+use anyhow::{Context, Result};
+use log::{debug, error};
+use nix::fcntl::OFlag;
+use nix::libc::c_int;
+use nix::sys::ptrace;
+use nix::sys::signal::{kill, raise, signal, Signal};
+use nix::sys::wait::{wait, WaitStatus};
+use nix::unistd::{fork, ForkResult, Pid};
+use nix::Error;
+use syscalls::x86_64::Sysno;
+
+const AT_FDCWD: u64 = 0xffffff9c;
+
+/// List of paths that are forbidden to modify.
+pub const FORBIDDEN_PATHS: &[&str] = &[
+  "/home/runner/workspace/.replit",
+  "/home/runner/workspace/replit.nix",
+  "/home/runner/workspace/.git/refs/replit/agent-ledger",
+];
+
+/// List of prefixes that are forbidden to delete.
+pub const FORBIDDEN_UNLINK_PREFIXES: &[&str] = &["/home/runner/workspace/.git/"];
+
+static mut CHILD_PID: Pid = Pid::from_raw(-1);
+
+/// Read a path (a NUL-terminated string) from the tracee.
+fn read_path(pid: Pid, mut addr: u64) -> Result<PathBuf> {
+  let mut buf = Vec::<u8>::with_capacity(1024);
+  let mut offset = (addr & 0x7) as usize;
+  addr -= offset as u64;
+  loop {
```

Comment:
Might be good to add some reasonable limits just in case this isn't a valid NUL terminated string